server/src/main/java/org/elasticsearch/repositories/RepositoryOperation.java

```diff
@@ -8,18 +8,69 @@
  */
 package org.elasticsearch.repositories;
 
+import org.elasticsearch.cluster.DiffableUtils;
+import org.elasticsearch.cluster.metadata.ProjectId;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.FixForMultiProject;
+
+import java.io.IOException;
+
 /**
  * Coordinates of an operation that modifies a repository, assuming that repository at a specific generation.
  */
 public interface RepositoryOperation {
 
+    /**
+     * Project for which repository belongs to.
+     */
+    @FixForMultiProject(description = "default implementation is temporary")
+    default ProjectId projectId() {
+        return ProjectId.DEFAULT;
+    }
+
     /**
      * Name of the repository affected.
      */
     String repository();
 
+    default ProjectRepo projectRepo() {
+        return new ProjectRepo(projectId(), repository());
+    }
+
     /**
      * The repository state id at the time the operation began.
      */
     long repositoryStateId();
+
+    record ProjectRepo(ProjectId projectId, String repoName) implements Writeable {
```

Comment:
It's pretty self-explanatory, but I think a small comment justifying why we need this record would be valuable.

Reply:
Sure see 474b4868e85fd6c82bf622da3f9e4abb46be51e9